src/page.rs

```diff
@@ -1,33 +1,35 @@
 use std::ffi::{CStr, CString};
 use std::io::Read;
-use std::ptr;
-use std::slice;
-
-use serde::{Deserialize, Serialize};
+use std::ptr::{self, NonNull};
 
 use mupdf_sys::*;
+use serde::{Deserialize, Serialize};
 
 use crate::{
-    context, Buffer, Colorspace, Cookie, Device, DisplayList, Error, Link, Matrix, Pixmap, Quad,
-    Rect, Separations, TextPage, TextPageOptions,
+    context, rust_vec_from_ffi_ptr, unsafe_impl_ffi_wrapper, Buffer, Colorspace, Cookie, Device,
+    DisplayList, Error, FFIWrapper, Link, Matrix, Pixmap, Quad, Rect, Separations, TextPage,
+    TextPageOptions,
 };
 
 #[derive(Debug)]
 pub struct Page {
-    pub(crate) inner: *mut fz_page,
+    /// Ideally we'd use `Unique` here to signify ownership but that's not stable
+    pub(crate) inner: NonNull<fz_page>,
     pub(crate) doc: *mut fz_document,
 }
 
+unsafe_impl_ffi_wrapper!(Page, fz_page, fz_drop_page);
+
 impl Page {
-    pub(crate) unsafe fn from_raw(raw: *mut fz_page) -> Self {
+    pub(crate) unsafe fn from_raw(raw: NonNull<fz_page>) -> Self {
```

Comment:
maybe this should be renamed to `from_non_null`? We could also consider keepping the `from_raw` method, make it call `from_non_null` internally and return a `Result` instead.

Reply:
Changed in 6c4d41e217e58aa269813470dd0b9e09ac1cc415